pages/blog/2020-03-12-nextjs-without-client-side-react.md

```diff
@@ -0,0 +1,164 @@
+### Custom Main component
+
+This is not really necceary but to satisfy my mild OCD I wanted to get rid of the `<div id="next">` wrapper around my content.
+
+```jsx
+import { Main } from "next/document";
+export class MyMain extends Main {
+  render() {
+    const { html } = this.context._documentProps;
+    return <div dangerouslySetInnerHTML={{ __html: html }} />;
+  }
+}
+```
+
+### Stitching it together: Custom Document
+
+Customizing the `pages/_document.js` is quite common, only we're going to do it slightly differently. The key here is that we omit the `<NextScripts/>` component and keep things as slimmed down as possible
+
+```jsx
+import Document from "next/document";
+export default class extends Document {
+  render() {
+    return (
+      <html lang="en">
+        <MyHead>
+          <meta
+            name="viewport"
+            content="width=device-width, initial-scale=1, shrink-to-fit=no"
+          />
+        </MyHead>
+        <body>
+          <MyMain />
+        </body>
+      </html>
+    );
+  }
+}
+```
+
+For brevity, I've left out the CSS-in-JS parts in the example above, I'll leave that for another article.
+
+### End Result
+
+With this bare bone setup we get the following HTML output:
+
+```html
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta
+      name="viewport"
+      content="width=device-width, initial-scale=1, shrink-to-fit=no"
+    />
+    <meta charset="utf-8" />
+    <title>webcloud</title>
+  </head>
+  <body>
+    <div>Welcome to my no JavaScript Next Page</div>
+  </body>
+</html>
+```
+
+The end result is pretty cool. I get a HTML page that is serve-side rendered with React, generated from a Markdown file. Things like code syntax highlighting is done on the server, so need need for a client-side runtime.
+
+I'm using [Fela](http://fela.js.org/) for CSS-in-JS so that the critical amount of CSS needed to render a specific page is inline directly in the head. This means a HTML page that is entirely self-contained, no external CSS or JavaScript files required, only a single request!
+
+
+
+### Bonus: Adding vanilla JavaScript back in the mix
+
+OK I lied, this page is not entirely JavaScript free. You may have noticed that I have a dark/light theme switching button in the top right corner. To achieve this I've added roughly 0.5kB of render blocking JavaScript that detects the users OS color preference, and modified CSS global properties before the page is rendered. I will write an article about that soon!
+
+
+### Bonus: Getting back that "SPA" feeling
+
+One thing that *is* nice with isomorphic/pages is that navigating between pages is more or less instant. This is not such a huge problem for me for most of my content is long form, and I don't expect my visitors to be doing a lot of navigation between pages.
+
+With that being said I've added a small script called [instant.page](https://instant.page/) that uses a `<link preload>` strategy to add a near "SPA-like" experience for non SPA websites.
```

Comment:
What about doing this without the instant.pge ?  why not just `link prefetch/preload`?